consensus/state_sync/state_sync.go

```diff
@@ -0,0 +1,104 @@
+
+	// `NetworkSend` function contract:
+	// - sends data to an address via P2P network
+	SetNetworkSend(NetworkSend func(data []byte, address cryptoPocket.Address) error)
+
+	// -- Sync modes --
+
+	// In the StateSync protocol, the Node fields valid BlockRequests from its peers to help them CatchUp to be Synced.
+	// This sub-protocol is continuous throughout the lifecycle of StateSync.
+	RunServerMode()
+
+	// In PacemakerMode, the Node is caught up to the latest block and relies on the Consensus Module’s Pacemaker
+	// to maintain a synchronous state with the global SyncState.
+	// - UpdatePeerMetadata from P2P module
+	// - UpdateSyncState
+	// - Defer Sync to Pacemaker
+	// - If `localSyncState.Height < globalNetworkSyncState.Height` -> RunSyncMode() // careful about race-conditions with Pacemaker
+	RunPacemakerMode()
+
+	// Runs sync mode 'service' that continuously runs while `localSyncState.Height < globalNetworkSyncState.Height`
+	// - UpdatePeerMetadata from P2P module
+	// - Retrieve missing blocks from peers
+	// - Process retrieved blocks
+	// - UpdateSyncState
+	// - If `localSyncState.Height == globalNetworkSyncState.Height` -> RunPacemakerMode()
+	RunSyncMode()
+
+	// Returns the `highest priority aka lowest height` missing block heights up to `max` heights
+	GetMissingBlockHeights(state SyncState, max int) (blockHeights []int64, err error)
+
+	// Random selection of eligilbe peers enables a fair distribution of blockRequests over time via law of large numbers
+	// An eligible peer is when `PeerMeta.MinHeight <= blockHeight <= PeerMeta.MaxHeight`
+	GetRandomEligiblePeersForHeight(blockHeight int64) (eligiblePeer PeerSyncMeta, err error)
+
+	// Uses `NetworkSend` to send a `BlockRequestMessage` to a specific peer
+	SendBlockRequest(peerId string) error
+
+	// Uses 'NetworkSend' to send a `BlockResponseMessage` to a specific peer
+	// This function is used in 'ServerMode()'
+	HandleBlockRequest(message BlockRequestMessage) error
+
+	// Uses `HandleBlock` to process retrieved blocks from peers
+	// Must update sync state using `SetMissingBlockHeight`
+	ProcessBlock(block *typesCons.Block) error
+}
+
+type SyncState interface {
+	// latest local height
+	LatestHeight() int64
+	// latest network height (from the aggregation of Peer Sync Meta)
+	LatestNetworkHeight() int64
+	// retrieve peer meta (actively updated through churn management)
+	GetPeers() []PeerSyncMeta
+	// returns ordered array of missing block heights
+	GetMissingBlockHeights() []int64
+	// set missing block height if an error occurred during processing
+	SetMissingBlockHeight(blockHeight int64) error
+}
+
+type BlockRequestMessage interface {
+	// the height the peer wants from the block store
+	GetRequestedHeight() int64
+}
+
+type BlockResponseMessage interface {
+	// the bytes of the requested block from the block store
+	GetBlockBytes() []byte
+}
+
+// TODO: needs to be shared between P2P as the Churn Manageent Process updates this information
```

Comment:
Can you create a ticket for this TODO?

I think understanding what part of the interface will be exposed under `shared/modules/p2p_module.go` is important since it'll affect @deblasis work on peer discovery & churn.

Reply:
https://github.com/pokt-network/pocket/issues/357